Firestore/Source/Model/FSTFieldValue.mm

```diff
@@ -89,7 +108,8 @@ - (FSTTypeOrder)typeOrder {
 
 - (id)value {
   return [self valueWithOptions:[FSTFieldValueOptions
-                                    optionsForSnapshotOptions:[FIRSnapshotOptions defaultOptions]]];
+                                    optionsForSnapshotOptions:[FIRSnapshotOptions defaultOptions]
+                                    timestampBehavior:FIRTimestampBehaviorDefault]];
```

Comment:
I really don't like this. It looks like this is the place in code where the default is actually set, regardless of what's inside `FIRFirestoreSettings`. I'm trying to work around having two places where default is set by introducing an "alias" for the default enum value. Let me know what you think.

Reply:
Am I correct in thinking that this code is only hit by tests (FIRDocumentSnapshot always uses the valueWithOptions: method instead)?  If so, then it seems like we should either:

1. Remove the method and update all tests to use valueWithOptions.
2. Use FIRTimestampBehaviorReturnTimestamp and update our tests to deal with it (since this will be the default in the future).
3. Leave this as FIRTimestampBehaviorReturnDate but put a comment explaining this is only used by tests and we should consider updating them to work with the new timestamp behavior.

Reply:
I think that would be an improvement, and I'd be fine with that approach.

Unfortunately I think FSTFieldValue is in a bit of a wonky state, especially on iOS. Note the comment on the `value` method:
```
 * Converts an FSTFieldValue into the value that users will see in document snapshots.
 *
 * TODO(mikelehen): This conversion should probably happen at the API level and right now `value` is
 * used inappropriately in the serializer implementation, etc.  We need to do some reworking.
 ```

Since `value` is meant to compute the _user-facing_ value, we shouldn't actually be using it in FSTRelationFilter or FSTSerializerBeta and expecting to get back a specific internal representation...

Instead we should directly expose the internal representation, e.g. in FSTFieldValue.h, add a property to FSTFieldValue:
```
@interface FSTReferenceValue : FSTFieldValue <FSTDocumentKey *>
...
@property(nonatomic, strong, readonly) FSTDocumentKey *internalValue;
@end
```

Several other FSTFieldValue subclasses already do this (search for internalValue in FSTFieldValue.h).  And then we could use `internalValue` from `FSTRelationFilter` instead of `value`. Applying the same rework to FSTSerializerBeta (and the tests?) we could ultimately get rid of the `value` method entirely. Internal code should use `internalValue` and the API layer code should use `valueWithOptions`.

And in the longer-term, I think we should actually get rid of `valueWithOptions` too, moving the "model types" => "API types" conversion in the API layer instead, which would clean up some other warts (like the fact that FSTReferencValue's `value*` method returns an FSTDocumentKey which then must be converted again into a FIRDocumentReference at the API layer).

All of this said, I know Zhi has already done FieldValue C++ porting work and I don't know how all of this overlaps... and we probably don't want to turn your PR into a "FieldValue revamp" PR. :-)  So feel free to pick whatever feels appropriate for your PR.  As I said at the beginning, your suggestion to flip the value / valueWithOptions relationship seems fine for now.

Reply:
I agree, and I was thinking along similar lines:
* `value` and `valueWithOptions` seem to be pretty different despite the name similarity - `value` is some sort of a "canonical" representation, and `valueWithOptions` is "user-facing";
* `DocumentSnapshot` might be a better place to apply the options. It will require type casting which isn't perfect, but having all `FieldValue`s implement `valueWithOptions` when in fact nearly all of them ignore options also isn't. I can work on that after Timestamp.

For the purposes of this PR, I did reverse the forwarding between `value` and `valueWithOptions`, and make `FSTSerializerBeta` call `[timestampValue value]` instead of `[(FSTTimestampValue*)timestampValue internalValue]`.